### ISSUE TITLE
Bug for Raw SQL queries for unmapped types not working

### ISSUE THREAD

**Issue by jamesperrin** [dotnet/efcore#33453]
There is a bug in the EF Core SqlQuery() method that's preventing code to work with unmapped types like DTOs (data transfer object).

The EF Core SqlQuery() method should function similarly to EF 6 System.Data.Entity SqlQuery() method. There are use cases for returning records from the database that aren't tracked by the context for data transfer objects and entity types. The method should support raw SQL and stored procedures.

When I try the EF Core SqlQuery() method using raw SQL or a stored procedure, I receive the following error messages for a know database record.

"Data is Null. This method or property cannot be called on Null values."

OR 

"Cannot create a DbSet for 'ContactResultDTO' because this type is not included in the model for the context."

OR

"'FromSql' or 'SqlQuery' was called with non-composable SQL and with a query composing over it. Consider calling 'AsEnumerable' after the method to perform the composition on the client side."

### Include your code

```C#
// DTO
public class ContactResultDTO
{
	public int Id { get; set; }
	public string LastName { get; set; }
	public string FirstName { get; set; }
	public string MiddleName { get; set; }
	public string Title { get; set; }
	public string Email { get; set; }
	public string Department { get; set; }
}

// Using raw SQL
public async Task<ContactResultDTO?> GetContactResultDtoAsync(int contactId)
{
	return await _appDbContext.Database
                .SqlQuery<ContactResultDTO>($"SELECT Id, LastName, FirstName, MiddleName, Title, Email, Department, FROM Contact WHERE Id = {contactId}").FirstOrDefaultAsync();
}

// Using stored procedure.
public async Task<ContactResultDTO?> GetContactResultDtoAsync(int contactId)
{
	var contactIdParameter = contactId <= 0
	? new SqlParameter() { ParameterName = "@ContactId", SqlDbType = SqlDbType.Int, Value = 0 }
	: new SqlParameter() { ParameterName = "@ContactId", SqlDbType = SqlDbType.Int, Value = contactId };

	var sqlParameters = new[]
	{
		contactIdParameter
	};
					
	return await _appDbContext.Database
	.SqlQuery<ContactResultDTO>("EXEC [dbo].[uspGetContact] @ContactId", sqlParameters)
	.FirstOrDefaultAsync();
}
```

### Include provider and version information

EF Core version: Microsoft.EntityFrameworkCore v8.0.3
Database provider: Microsoft.EntityFrameworkCore.SqlServer v8.0.3
Target framework: NET 8.0
Operating system: Windows 10 Enterprise
IDE: Visual Studio 2022 17.9.5

### References:

Entity Framework Core -  Microsoft.EntityFrameworkCore. RelationalDatabaseFacadeExtensions SqlQuery()
[https://github.com/dotnet/efcore/blob/5651f40fc2528d8e0cea312009e9d79a40a6c01c/src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs#L404](https://github.com/dotnet/efcore/blob/5651f40fc2528d8e0cea312009e9d79a40a6c01c/src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs#L404)

Entity Framework 6 - System.Data.Entity SqlQuery()
[https://github.com/dotnet/ef6/blob/main/src/EntityFramework/Database.cs#L523](https://github.com/dotnet/ef6/blob/main/src/EntityFramework/Database.cs#L523)

```C#
/// <summary>
/// Creates a raw SQL query that will return elements of the given generic type.
/// The type can be any type that has properties that match the names of the columns returned
/// from the query, or can be a simple primitive type.  The type does not have to be an
/// entity type. The results of this query are never tracked by the context even if the
/// type of object returned is an entity type.  Use the <see cref="DbSet{TEntity}.SqlQuery" />
/// method to return entities that are tracked by the context.
///
/// As with any API that accepts SQL it is important to parameterize any user input to protect against a SQL injection attack. You can include parameter place holders in the SQL query string and then supply parameter values as additional arguments. Any parameter values you supply will automatically be converted to a DbParameter.
/// context.Database.SqlQuery&lt;Post&gt;("SELECT * FROM dbo.Posts WHERE Author = @p0", userSuppliedAuthor);
/// Alternatively, you can also construct a DbParameter and supply it to SqlQuery. This allows you to use named parameters in the SQL query string.
/// context.Database.SqlQuery&lt;Post&gt;("SELECT * FROM dbo.Posts WHERE Author = @author", new SqlParameter("@author", userSuppliedAuthor));
/// </summary>
/// <typeparam name="TElement"> The type of object returned by the query. </typeparam>
/// <param name="sql"> The SQL query string. </param>
/// <param name="parameters"> 
/// The parameters to apply to the SQL query string. If output parameters are used, their values will 
/// not be available until the results have been read completely. This is due to the underlying behavior 
/// of DbDataReader, see http://go.microsoft.com/fwlink/?LinkID=398589 for more details.
/// </param>
/// <returns>
/// A <see cref="DbRawSqlQuery{TElement}" /> object that will execute the query when it is enumerated.
/// </returns>
public DbRawSqlQuery<TElement> SqlQuery<TElement>(string sql, params object[] parameters)
{
	Check.NotEmpty(sql, "sql");
	Check.NotNull(parameters, "parameters");

	return
		new DbRawSqlQuery<TElement>(
			new InternalSqlNonSetQuery(_internalContext, typeof(TElement), sql, parameters));
}
```

**Comment by roji** [comment]
@jamesperrin please see below for some specific comments, but in general we need a minimal, runnable code sample rather than partial snippets as above.

> "Data is Null. This method or property cannot be called on Null values."

This is likely because your DTO type has some non-nullable property but where the database has a null value. You have to make sure that your DTO type corresponds to your database schema.

Specifically regarding the error you get when trying to invoke a stored procedure:

> "'FromSql' or 'SqlQuery' was called with non-composable SQL and with a query composing over it. Consider calling 'AsEnumerable' after the method to perform the composition on the client side."

On this, see the following note in our [docs](https://learn.microsoft.com/en-us/ef/core/querying/sql-queries#including-related-data):

> SQL Server doesn't allow composing over stored procedure calls, so any attempt to apply additional query operators to such a call will result in invalid SQL. Use [AsEnumerable](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.asenumerable) or [AsAsyncEnumerable](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.entityframeworkqueryableextensions.asasyncenumerable) right after [FromSql](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationalqueryableextensions.fromsql) or [FromSqlRaw](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationalqueryableextensions.fromsqlraw) to make sure that EF Core doesn't try to compose over a stored procedure.

In other words, the problem is with your attempted composition of FirstOrDefaultAsync over a stored procedure call - that isn't supported by SQL Server (nothing to do with EF).

**Comment by jamesperrin** [comment]
@roji You are correct. The methods SqlQuery and SqlQueryRaw return an IQueryable which is an implementation of System.Collections.Generic.IEnumerable, basically a collection of objects. I updated my code accordingly.

```C#
public async Task<ContactResultDTO?> GetContactResultDtoAsync(int contactId)
{
	var contactIdParameter = contactId <= 0
		? new SqlParameter() { ParameterName = "@ContactId", SqlDbType = SqlDbType.Int, Value = 0 }
		: new SqlParameter() { ParameterName = "@ContactId", SqlDbType = SqlDbType.Int, Value = contactId };

	var sqlParameters = new[]
	{
		contactIdParameter
	};

	var contact = await _appDbContext
		.Database
		.SqlQueryRaw<ContactResultDTO>("EXECUTE [dbo].[uspGetContact] @ContactId", sqlParameters)
		.ToListAsync();

	return contact.FirstOrDefault();
}
```
